RBD: Fix encrypted restore with different kms config

Rakshith-R opened the pull request:
- rbd: modify copyEncryptionConfig to accept copyOnlyPassphrase arg 

`During PVC snapshot/clone both kms config and passphrase needs to copied,
while for PVC restore only passphrase needs to be copied to dest rbdvol
since destination storageclass may have another kms config.`

- e2e: modify validatePVCSnapshot() to use restoreSCName & restoreKMS

- e2e: add testcase for PVC restore from vaultKMS to vaultTenantSAKMS 
- e2e: add test for PVC restore from vaultKMS to userSecretMetadataKMS


Example of failed e2e run when vaultTenantSAKMS is used as restore SC 
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.20/detail/mini-e2e_k8s-1.20/2627/pipeline

search for `secret ID` , mount fails because it can't find encryption passphrase.

Rakshith-R committed:
rbd: modify copyEncryptionConfig to accept copyOnlyPassphrase arg

During PVC snapshot/clone both kms config and passphrase needs to copied,
while for PVC restore only passphrase needs to be copied to dest rbdvol
since destination storageclass may have another kms config.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
e2e: modify validatePVCSnapshot() to use restoreSCName & restoreKMS

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
e2e: add testcase for PVC restore from vaultKMS to vaultTenantSAKMS

Signed-off-by: Rakshith R <rar@redhat.com>

Madhu-1 commented on internal/rbd/encryption.go:
```diff
@@ -135,8 +135,11 @@ func (ri *rbdImage) copyEncryptionConfig(cp *rbdImage) error {
 	}
-
-	cp.encryption, err = util.NewVolumeEncryption(ri.encryption.GetID(), ri.encryption.KMS)
+	if copyOnlyPassphrase {
+		cp.encryption, err = util.NewVolumeEncryption(cp.encryption.GetID(), cp.encryption.KMS)
+	} else {
+		cp.encryption, err = util.NewVolumeEncryption(ri.encryption.GetID(), ri.encryption.KMS)
+	}
```
instead of this new variable, the same can be done if we compare KMS/ID?

Rakshith-R replied:
Actually cp.encrypyion is `nil` at places, so we cannot do that.
(Since it is reused for snapshotting and cloning (+intermediate steps))
And to have very minimal changes to be able to backport to 3.4.

Rakshith-R commented:
/retest ci/centos/mini-e2e-helm/k8s-1.20

Rakshith-R commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.20

failed due to #2264 
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.20/detail/mini-e2e-helm_k8s-1.20/2627/pipeline

nixpanic commented on internal/rbd/encryption.go:
```diff
@@ -135,8 +135,11 @@ func (ri *rbdImage) copyEncryptionConfig(cp *rbdImage) error {
-
-	cp.encryption, err = util.NewVolumeEncryption(ri.encryption.GetID(), ri.encryption.KMS)
+	if copyOnlyPassphrase {
+		cp.encryption, err = util.NewVolumeEncryption(cp.encryption.GetID(), cp.encryption.KMS)
```
this seems wrong, setting `cp.encryption` based on the contents of `cp.encryption` is weird. It would need at minimum a very clear comment explaining why this is needed, and works at all.

Rakshith-R replied:
>this seems wrong, setting cp.encryption based on the contents of cp.encryption is weird. It would need at minimum a very clear comment explaining why this is needed, and works at all.


yeah, it was redundant :facepalm: , fixed now, PTAL
destination rbdvol.encryption for snapshot restore PVC will be set/Initialized earlier, so its not required

nixpanic commented on e2e/utils.go:
```diff
@@ -867,6 +871,29 @@ func validatePVCSnapshot(
+						// check new passphrase created
+						stdOut, stdErr := restoreKMS.getPassphrase(f, imageData.csiVolumeHandle)
+						if stdOut != "" {
+							e2elog.Logf("successfully read the passphrase from vault: %s", stdOut)
```
this refers to "vault", but that is not the only passphrase storage that cam be configured? Is the error of `getPassphrase()` not clear enough?

Rakshith-R replied:
>this refers to "vault", but that is not the only passphrase storage that cam be configured? Is the error of getPassphrase() not clear enough?

The stdOut is not required here, removed it.

nixpanic commented on e2e/rbd.go:
```diff
@@ -1145,6 +1145,63 @@ var _ = Describe("RBD", func() {
+				restoreSCName := "restore-sc"
+				err := deleteResource(rbdExamplePath + "storageclass.yaml")
+				if err != nil {
+					e2elog.Failf("failed to delete storageclass with error %v", err)
```
new error logging should ideally not include the repetitive/obvious `with error`, but only `<message>: err`.

Rakshith-R replied:
done

Rakshith-R committed:
e2e: add nolint:param to retryKubectlArgs

Currently only kubectlCreate arg is used with retryKubectlArgs(),
But it maybe used later on.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
rbd: copyEncryptionConfig for thickProvisioned snap restore too

This commit adds bugfix to copy encryption passphrase for thick
provisioned PVC restored from snapshot.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
e2e: add testcase for thick encrypted PVC restore

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

nixpanic commented on internal/rbd/encryption.go:
```diff
@@ -123,7 +123,7 @@ func (ri *rbdImage) setupEncryption(ctx context.Context) error {
 // from the original, so that both encrypted passphrases (potentially, depends
 // on the DEKStore) have different contents.
-func (ri *rbdImage) copyEncryptionConfig(cp *rbdImage) error {
+func (ri *rbdImage) copyEncryptionConfig(cp *rbdImage, copyOnlyPassphrase bool) error {
```
it would help to explain in the above comment when to set `copyOnlyPassphrase` to `true`, and what it exactly does.

nixpanic commented on internal/rbd/encryption.go:
```diff
@@ -135,8 +135,9 @@ func (ri *rbdImage) copyEncryptionConfig(cp *rbdImage) error {
 	}
-
-	cp.encryption, err = util.NewVolumeEncryption(ri.encryption.GetID(), ri.encryption.KMS)
+	if !copyOnlyPassphrase {
```
add empty lines around this if-block to increase readability?

Move the error check and `SetDEKStore` inside the if-statement too?

nixpanic commented on e2e/utils.go:
```diff
@@ -1353,6 +1353,7 @@ func retryKubectlFile(namespace string, action kubectlAction, filename string, t
 // retryKubectlArgs takes a namespace and action telling kubectl what to do
 // with the passed arguments. This function retries until no error occurred, or
 // the timeout passed.
+// nolint:unparam // retryKubectlArgs will be used with kubectlDelete arg later on.
```
should `deleteResource` use this function then?

Rakshith-R replied:
its using `retryKubectlInput`  
https://github.com/ceph/ceph-csi/blob/90d246fc55142fbcfe9ef8fd34c4a20526c00ce0/e2e/utils.go#L126

nixpanic reviewed (commented):
just a few nits, looks pretty complete to me

Rakshith-R commented:
/retest ci/centos/upgrade-tests-rbd

Rakshith-R commented:
> /retest ci/centos/upgrade-tests-rbd

failed due to #2264 

https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/upgrade-tests-rbd/detail/upgrade-tests-rbd/3258/pipeline
